crates/sui-types/src/temporary_store.rs

```diff
@@ -832,72 +683,292 @@ impl<S: ObjectStore> TemporaryStore<S> {
+                    total_input_sui += self.get_input_sui(id, *input_version)?;
+                    // else, the wrapped object was either:
+                    // 1. freshly created, which means it has 0 contribution to input SUI
+                    // 2. unwrapped from another object A, which means its contribution to input SUI will be captured by looking at A
+                }
+                DeleteKind::UnwrapThenDelete => {
+                    // an unwrapped option was wrapped in input object or dynamic field A, which means its contribution to input SUI will
+                    // be captured by looking at A
+                }
+            }
         }
-        let gas_summary = &self.gas_charged.as_ref().unwrap().1;
+
+        // we do account for the "storage rebate inflow" (portion of the storage rebate which flows back into the storage fund).
+        let gas_summary = &self
+            .gas_charged
+            .as_ref()
+            .ok_or(ExecutionError::invariant_violation(
+                "Failed unwrapping gas_charged in SUI conservation checking",
+            ))?
+            .1;
         let storage_fund_rebate_inflow =
             gas_summary.storage_fund_rebate_inflow(self.storage_rebate_rate);
 
-        // total SUI in input objects
-        let input_sui = self.mutable_input_refs.iter().fold(0, |acc, o| {
-            acc + self
-                .input_objects
-                .get(&o.0)
-                .unwrap()
-                .get_total_sui(&self)
-                .unwrap()
-        });
-        // if a dynamic field object O is written by this tx, count get_total_sui(pre_tx_value(O)) as part of input_sui
-        let dynamic_field_input_sui = self
-            .dynamic_fields_touched_do_not_call_before_fixing()
-            .iter()
-            .fold(0, |acc, id| {
-                acc + self
-                    .store
-                    .get_object(id)
-                    .unwrap()
-                    .unwrap()
-                    .get_total_sui(&self)
-                    .unwrap()
-            });
-        // sum of the storage rebate fields of all objects written by this tx
-        let mut output_rebate_amount = 0;
-        // total SUI in output objects
-        let output_sui = self.written.values().fold(0, |acc, (o, _)| {
-            output_rebate_amount += o.storage_rebate;
-            acc + o.get_total_sui(&self).unwrap()
-        });
-
         // storage gas cost should be equal to total rebates of mutated objects + storage fund rebate inflow (see below).
         // note: each mutated object O of size N bytes is assessed a storage cost of N * storage_price bytes, but also
         // has O.storage_rebate credited to the tx storage rebate.
-        // TODO: figure out what's wrong with this check. The one below is more important, so going without it for now
+        // TODO: figure out what's wrong with this check. The one below is more important, but this should still hold.
```

Comment:
@dariorussi This would be a good thing to try to enable at the same time as your changes to the meaning of `storage_rebate` vs `storage_cost`--there are only a few tests that fail, and the discrepancy is always very small (about the size of `storage_rebate_inflow`)

Reply:
will do and of course we'll talk